src/CompositeRheologies.jl

```diff
@@ -0,0 +1,794 @@
+end
+
+
+
+"""
+    local_iterations_εII(c::CompositeRheology{T,N}, εII_total, args)
+
+This performs nonlinear Newton iterations for `τII` with given `εII_total` for cases where we have both serial and parallel elements.
+"""
+@inline function local_iterations_εII(
+    c::CompositeRheology{T,N,Npar, is_par}, 
+    εII_total::_T, 
+    args; 
+    tol=1e-6, 
+    verbose=false, τ_initial=nothing, ε_init=nothing
+) where {T,N,Npar,is_par, _T}
+    
+    # Compute residual
+    n = Npar+1;             # total size of unknowns
+    x = zero(εII_total)
+    
+    # Initial guess of stress & strainrate
+    if isnothing(τ_initial)
+        τ_initial = compute_τII_harmonic(c, εII_total, args)
+    end
+
+    verbose && println("τII guess = $τ_initial")
+
+    x    = @MVector ones(_T, n)
+    x   .= εII_total
+    x[1] = τ_initial
+
+    j = 1;
+    for i=1:N
+        if is_par[i]
+            j += 1
+            x[j] = compute_εII_harmonic_i(c, τ_initial, args,i)   
+        end
+    end
+    
+    r = @MVector zeros(_T,n);
+    J = @MMatrix zeros(_T, Npar+1,Npar+1)   # size depends on # of parallel objects (+ likely plastic elements)
+    
+    # Local Iterations
+    iter = 0
+    ϵ = 2 * tol
+    τII_prev = τ_initial
+    τ_parallel = _T(0)
+    max_iter = 1000
+    while (ϵ > tol) && (iter < max_iter)
+        iter += 1
+
+        τ   = x[1]
+  
+        # Update part of jacobian related to serial elements
+        r[1]   = εII_total - compute_εII_elements(c,τ,args)
+        J[1,1] = dεII_dτII_elements(c,x[1],args);
+        
+        # Add contributions from || elements
+        fill_J_parallel!(J, r, x, c, τ, args)
+        
+        #=
+        j=1;
+        for i=1:N
+            
+            if is_par[i]
+                j += 1
+                εII_p = x[j]
+                r[1] -= εII_p
+                τ_parallel = compute_τII_i(c, εII_p, args, i)        
+                r[j]       =  (τ - τ_parallel);                              # residual (stress should be equal)
+                J[j,j]     = -dτII_dεII_i(c, εII_p, args, i)    
+                
+                J[j,1]     =  1.0
+                J[1,j]     =  1.0
+            end
+        end
+        =#
+
+        # update solution
+        dx  = J\r 
+        x .+= dx   
+
+        ϵ    = sum(abs.(dx)./(abs.(x)))
+        verbose && println(" iter $(iter) $ϵ")
+    end
+    verbose && println("---")
+    
+    if (iter == max_iter)
+        error("iterations did not converge")
+    end
+
+    τII = x[1]
+
+    return τII
+end
+
+# Helper function
+@generated function fill_J_parallel!(J, r, x, c::CompositeRheology{T, N, Npar, is_par, τ_it, P_it, λ_it}, τ, args) where {T, N, Npar, is_par, τ_it, P_it, λ_it}
+    quote
+        Base.@_inline_meta
+        j = 1
+        Base.Cartesian.@nexprs $N i -> j = @inbounds _fill_J_parallel!(J, r, x, c.elements[i], τ, args, $(is_par)[i], j)
+        return nothing
+    end
+end
+
+function _fill_J_parallel!(J, r, x, elements, τ, args, is_par, j)
```

Comment:
This is allocating because the compiler is failing to inline this function.  Adding the `@inline` macro here removes all the allocations.